lib/streamlit/dataframe_util.py

```diff
@@ -686,9 +686,9 @@ def convert_arrow_bytes_to_pandas_df(source: bytes) -> DataFrame:
 def _show_data_information(msg: str) -> None:
     """Show a message to the user with important information
     about the processed dataset."""
-    from streamlit.delta_generator import main_dg
+    from streamlit.delta_generator_singletons import get_main_dg
 
-    main_dg.caption(msg)
+    get_main_dg().caption(msg)
```

Comment:
I believe we already discussed this somewhere, but is `main_dg` the same as the active_dg and if so, shouldn't we rename it to `active_dg`. Otherwise, its a bit confusing, since in this case the caption shoul be added to the sidebar if the dataframe is used in the sidebar and not to the main container.

Reply:
No, the `main_dg` and the `active_dg` are not the same. The active_dg based on [this function def](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L294-L309) differentiates between the invocation type (`with st.sidebar(): st.write()` vs `st.sidebar.write()`).

You probably mean the `main_dg` and the `DeltaGenerator._main_dg` [property](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L311-L317). Theoretically they are also not the same, because the `_main_dg` property recursively goes up the chain to find in which root a dg is. However, in practice this _should_ be the same as the root of all delta generators should be the `main_dg`. I agree that if this hypothesis is correct, we should clean it up and simplify. But I don't want to touch this part of the legacy code yet and introduce potentially functional-changing code changes in this mainly refactorial PR. However, it is on our list to clean up in the next phase of the project!

Reply:
But I think what we need in most situations is actually the active dg and not the main dg. E.g. the caption above is broken if you use it with: `st.sidebar.dataframe(...)`

Can we also expose the active dg as a property of the singelton?

Reply:
I think its also confusing / inconsistent that that both of these statements put the element into the sidebar:
```
with get_dg_singleton_instance().sidebar_dg:
    get_dg_singleton_instance().main_dg.caption("Foo")

with get_dg_singleton_instance().main_dg:
    get_dg_singleton_instance().sidebar_dg.caption("Bar")
```

Reply:
I think that's general feedback about the library and not something we should address in this PR. The statement above is the equivalent of 

```python
with st._sidebar_dg:
  st._main_dg.caption("Foo")

with st._main_dg:
  st._sidebar_dg.caption("Bar')
```
which is how the library works.
Maybe the confusing part previously was that the `st._main_dg` was equivalent to the delta_generator.py's `main_dg` variable, not `DeltaGenerator._main_dg`.


WIth regards to:
> But I think what we need in most situations is actually the active dg and not the main dg. E.g. the caption above is broken if you use it with: st.sidebar.dataframe(...)
>
> Can we also expose the active dg as a property of the singelton?


I am not sure that is true. The changed line of code used the `main_dg`, like it is using it now. If you think that has always been wrong, then we should discuss it, but should come in a different PR and after thoroughly thinking about it.

Reply:
> I am not sure that is true. The changed line of code used the main_dg, like it is using it now. If you think that has always been wrong, then we should discuss it, but should come in a different PR and after thoroughly thinking about it.

Yep, my recent change from `st.caption()` to `main_dg.caption()` introduced this issue. So, it's currently broken in `develop`. But to properly fix it, we should expose the `active_dg` as well since `main_dg` is not the correct thing to use here.

Reply:
> Maybe the confusing part previously was that the st._main_dg was equivalent to the delta_generator.py's main_dg variable 

Yep, that's definitely a problem to address. Can be in a follow-up PR. 

> I think that's general feedback about the library and not something we should address in this PR.

Yep, doesn't need to be in this PR. But I think it would be good to address this as part of the overall refactoring project.

Reply:
> Yep, my recent change from st.caption() to main_dg.caption() introduced this issue.

Hmhm but I think `st.caption()` and your change of `main_dg.caption` is equivalent too!
Because in the `__init__.py` we assign `caption = st._main_dg` and `st._main_dg` is equal to delta_generator.main_dg. So your change uses the delta_generator module directly (which I think is good) but other than that it did not change the behavior.

Reply:
hmm, interesting. Is there a way to add another element from within an element to the active delta generator location?

Reply:
It does not seem to be the case at that we can properly support both of these uses :(

```python
with st.sidebar:
    st.dataframe(BASE_TYPES_DF)

st.sidebar.dataframe(BASE_TYPES_DF)
```

<img width="832" alt="image" src="https://github.com/user-attachments/assets/8d1cfd93-dab6-422b-8ffb-a965b53a5a0c">